### ISSUE TITLE
[APP] Enedis

### ISSUE THREAD

**Issue by LucBerge** [PipedreamHQ/pipedream#4112]
**Name of app / service**
Enedis

**Link to developer documentation**
[Data-connect](https://datahub-enedis.fr/data-connect/)
[Official documentation](https://datahub-enedis.fr/data-connect/documentation/metering-data-v4/)

**Is lack of support preventing you from building workflows, or do you have a workaround?**
No

**Are there specific actions, or triggers, you'd like to see for this app? Please let us know here or use the Action and Trigger issue templates to open requests for each!**

Action:
- Get daily consumption
- Get load curve
- Get max power
- Get daily production
- Get production load curve
- Get contracts
- Get adresses
- Get identity
- Get contact

**Comment by dylburger** [comment]
Thanks for the request! Did you know that anyone can [author and publish their own Pipedream actions](https://pipedream.com/docs/components/quickstart/nodejs/actions/) and [sources](https://pipedream.com/docs/components/quickstart/nodejs/sources/)? Would you be interested in helping out with that? We're happy to answer any questions or support you in any way.

If you don't have the time, would you be willing to reach out to the app provider to see if they're interested in building these? When we work with an app partner, it improves the quality of the integration, and it helps when this kind of request comes from a customer like you.

**Comment by dylburger** [comment]
Thanks for the request! We've added this to [our integrations backlog](https://github.com/orgs/PipedreamHQ/projects/8/views/1), and we'll update this ticket with progress.

**Comment by LucBerge** [comment]
@sergioeliot2039 here is the [full API doc](https://datahub-enedis.fr/data-connect/documentation/metering-data-v4/) but you need an account.

**Comment by LucBerge** [comment]
The pipeream application string is `e847028b-6be7-4397-8e42-82013595a9b8`

**Comment by dannyroosevelt** [comment]
Thanks @LucBerge I can take a look at this, but we'll need a `client_secret` for the Pipedream OAuth app as well (which should not be pasted here on this issue).

**Comment by dannyroosevelt** [comment]
FYI I also just submitted a request to create an OAuth app.

**Comment by LucBerge** [comment]
Hello,
This is an update on the work in progress for this APP.

Great work have been made thanks to @dannyroosevelt. Integrating this app is a mess of administrative tasks !

The APP is currently in `sandbox mode` which means the data we can access are fake. The purpose is to validate the code while wainting for the APP to go `production mode`.

To go `production mode` pipedream need to fill [this form](https://datahub-enedis.fr/renseigner-une-entite/) in order to be granted to access real data. As a US based company, you must select `Autre` for the field `Type entité`. In step 1 of 3, they are asking for any document which can be used as the ID of the company.

**Comment by LucBerge** [comment]
### BUG

For now, I have a troubles to connect my account in `sandbox mode`:

On my workflow, when I try to connect my account (1), a connection window is open (2). The redirection to pipedream is not working. Pipedream support already have acces to my workflow : https://pipedream.com/@LucBerge/test-enedis-p_q6CnmyB

![](https://www.pixenli.com/image/VcySiF0Z)

The url is the following
`https://gw.hml.api.enedis.fr/dataconnect/v1/oauth2/authorize?client_id=c8a45b5b-789c-4987-8d85-fb00a427bf08&duration=P6M&redirect_uri=https%3A%2F%2Fapi.pipedream.com%2Fconnect%2Foauth%2Foa_gK1i37%2Fcallback&response_type=code&scope=&state=dV9yOWhiYlZnfDdkMGIzMWRlMGYwOTcxMmFkOTU3ZDhhZTRiYTEzYzdl`

If I copy the link and open it in a regular browser tab, the redirection works but I have the following pipedream error:

> Pipedream failed to connect to your account.
> 
> Please reach out to integration-errors@pipedream.com and provide this error ID: e913d677f0634b1db10dc7aea4c314ed

Strange behaviour going on here

**Comment by LucBerge** [comment]
### BUG

A different bug occurs with Microsoft Edge: The main page is redirected will the small one no. [Here is a video](https://gfycat.com/competentcrispantarcticgiantpetrel)

![](https://www.pixenli.com/image/myd13bpb)

The error code is `e35a01029c004a36a00a4e603501d2f2`

**Comment by LucBerge** [comment]
Hello, any news about this bug?

**Comment by dannyroosevelt** [comment]
Hey @LucBerge apologies for the delay getting back to you. Troubleshooting anything with the Enedis app is difficult since their servers are only available 9-5 Paris time, and the majority of our team is in US Pacific time, so we cannot test or debug anywhere close to our working hours.

FYI I am still going back and forth with their team re: approving our app.

**Comment by LucBerge** [comment]
I don't understand why you say their servers are only available 9-5. Where did you find this info ? I don't have troubles to connect (at 1am) with a python script I've made (see slack)

**Comment by dannyroosevelt** [comment]
Maybe this was only re: creating the application? But when I test their Sandbox Authorization endpoint right now I get a 503 Service Unavailable error.

![Screenshot by Dropbox Capture](https://user-images.githubusercontent.com/20780753/204934177-cd89fe64-0ad2-491e-b821-86413a8420ff.png)

![Screenshot by Dropbox Capture](https://user-images.githubusercontent.com/20780753/204935042-5c009896-fbdc-44cc-8ff0-f697fdaffb6f.png)

**Comment by dannyroosevelt** [comment]
That's a good idea, but I really need to be able to debug while testing. By the way I made a slight change to the OAuth Authorization URL yesterday (but I just can't validate that it made any difference...)

**Comment by dannyroosevelt** [comment]
@LucBerge unfortunately we're blocked on this again -- they approved our corporate verification, but they require we sign a contract to proceed and they are not able to share in English, which is a blocker for us right now.
![Screen Shot 2022-12-02 at 9 33 32 AM](https://user-images.githubusercontent.com/20780753/205351844-b48bdf6f-9235-4042-b178-392e429590c3.png)

**Comment by sergioeliot2039** [comment]
@dylburger jumping in with some input and digress on related language barrier when it comes to Project App Store

Perhaps this is a case where it could be helpful to reach out for a "sworn translation" of the contract. for this it is needed to reach a French to English translator with a government issued certification on translation document for legal purposes (a "sworn translator"). I totally get this is an unusual Pipedream activity but could be helpful for the matter of make sure the contract is signed while the appropriate understanding of its content.  Not very time efficient, but better than Pipedream being on the need to have legal representation in France to sign that.

Now, I recently requested API access to a service that is based in France too. I sent then over request in French and they denied, and wait for it, invited to create an account in  another service! 😅

![image](https://user-images.githubusercontent.com/25184117/205361755-27077009-c3cb-490b-826f-4d22dfce443f.png)

"we are sorry, but this decision is final, we invite to create an account in another service..." 

That said, I know there is Google Translation to get what the contract says in English, but maybe I can give faith on what is says (and probably @LucBerge ), if that's of helpful. My two cents

**Comment by LucBerge** [comment]
> they require we sign a contract to proceed and they are not able to share in English

...

> That said, I know there is Google Translation to get what the contract says in English, but maybe I can give faith on what is says (and probably @LucBerge too ), if that's of helpful. My two cents

If I can help you in anyway, just ask. I can translate !

**Comment by LucBerge** [comment]
@dannyroosevelt I'm able to retrieve daily consumption with sandbox endpoint on my workflow: https://pipedream.com/@LucBerge/test-enedis-p_q6CnmyB.

To make it work correctly, you will need to change some params in the requests:

**In step 1**
- `client_id` must be replaced by your public key [which you can find here](https://datahub-enedis.fr/mon-compte-tableau-de-bord/mon-compte-applications/)
- `redirect_uri` must be replaced by your redirection url `https://api.pipedream.com/connect/oauth/Foa_gK1i37/callback`
- `state` must be replaced by whatever you want
- `duration` can be up to `P3Y`

**In step 2**
- This is a small hack
- `code`, `state` and `usage_point_id` are normally retrieved on pipedream side once the redirection in step 1 is complete.

**In step 3**
- `redirect_uri` must be replaced by your redirection url `https://api.pipedream.com/connect/oauth/Foa_gK1i37/callback`
- `client_id` must be replaced by your public key [which you can find here](https://datahub-enedis.fr/mon-compte-tableau-de-bord/mon-compte-applications/)
- `secret_id` must be replaced by your secret key [which you can find here](https://datahub-enedis.fr/mon-compte-tableau-de-bord/mon-compte-applications/)
- 'code' must be replaced by the code received in step 2

Once step1, 2 and 3 are complete, the Bearer token received in step 3 is available 3 years as requested in step 1. It means the following step 4 can be done as much as you need without requesting a new token.

**Comment by dannyroosevelt** [comment]
Thanks for the prototype! 

1. The Authorization URL is not responding right now -- can you try on your side? See below screenshot
2. I just made an update to the app configuration on our end, which will permit you to enter your own Client ID and Secret. This will help bypass the current contractual obligations that are slowing down the app release for the time being, since I think your OAuth app has been approved for production, is that right?

![Screenshot by Dropbox Capture](https://user-images.githubusercontent.com/20780753/206255433-994f1593-8529-4fe5-9063-fa410ce3e0c5.png)

**Comment by LucBerge** [comment]
No my app cannot go production because I cannot sign the contract as a single individual. Thank you, I'll try tomorrow. If it's working I'll PR the actions

**Comment by LucBerge** [comment]
When I try to connect my Enedis account with my `client_id` and `client_secret` a window pops-up (while it should not with your bypass) with a JSON response

![](https://www.pixenli.com/image/YFkUq0T_)

**Comment by LucBerge** [comment]
Yes, the window in the bottom opened automatically when clicking continue

**Comment by dannyroosevelt** [comment]
Interesting. This is what I see when I test in Firefox:
![Screenshot by Dropbox Capture](https://user-images.githubusercontent.com/20780753/206525279-5dfcfe3b-8969-4283-9429-7a0e9272d121.png)

So you do not get that standard OAuth popup window?

**Comment by LucBerge** [comment]
I have this too now, because the server is off

**Comment by LucBerge** [comment]
Note that [they just released a new API](https://datahub-enedis.fr/data-connect/documentation/autorisation-v1/).
The endpoint we are using are still available.

Something interesting: Regarding [this page](https://datahub-enedis.fr/data-connect/ressources/decouvrir/) and [this one](https://datahub-enedis.fr/data-connect/ressources/migrer-votre-application/) which (has been updated), we no longer need the authorization code in sandbox

**Comment by dannyroosevelt** [comment]
Now that we've modified the app's configuration to use your own client ID and client secret, there shouldn't be any changes required for Pipedream in terms of creating a new application.

Let me know when you want to update the Authorization and Access Token URLs and we can do that.

**Comment by LucBerge** [comment]
@dannyroosevelt Yes you have some changes to do:

**In sandbox mode:**
You don't need to call the authorization endpoint anymore (sandbox mode only).
You need to change the token request:

POST https://ext.hml.api.enedis.fr/oauth2/v3/token
 
Params
```
redirect_uri: https://api.pipedream.com/connect/oauth/oa_gK1i37/callback
```

Body (application/x-www-form-urlencoded)
```
grant_type: client_credentials
client_id: <clinet_id>
secret_id: <secret_id>
```

**Comment by dannyroosevelt** [comment]
Got it, thanks. I just made the changes to the OAuth configuration on our end -- let me know how that works for you when you get your new app configured!

**Comment by LucBerge** [comment]
Not working. A secure window is opened with the following url:
> https://api.pipedream.com/connect/oauth/oa_gK1i37?app_id=app_13GhwZ&&&connect_start_ms=1670913562005&cfmap[client_id]=5750d3d9-2542-4a55-a5b5-10ee0f095350&cfmap[client_secret]=mysecretkey

The response is as follow:
>Pipedream failed to connect to your account.
> Please reach out to integration-errors@pipedream.com and provide this error ID: f7a6234f230f4b0bb06f21e4c14a3145

I noticed a weird thing in the url:
Between the first and second parameter, there is **&&&**. I'm not sure it is the problem

**Comment by LucBerge** [comment]
> Do you know if there is a way to programmatically retrieve the usage_points_id?

With the new API, **yes**: https://datahub-enedis.fr/data-connect/ressources/production/
But it's only available in production.

> Obtenir les numéros de PRM du client
>
> Une fois que le client a accepté de partager ses données, le navigateur le redirigera vers la page de votre choix, celle indiquée comme page de redirection au moment de la création de votre application. Dans l’URL se trouve les numéros de PRM que vous devez récupérer pour la suite des opérations.
>
> Le state est renvoyé avec la même valeur que vous aviez envoyée. Un code d’autorisation est également transmis : vous pouvez ignorer cette valeur.
>
> Exemple de réponse:
> `https://redirect.com/redirect?state=XYZ&usage_points_id=xxxxx,yyyyyy&code=134567281`

**Comment by LucBerge** [comment]
Do you need help to translate the contract ?

**Comment by dannyroosevelt** [comment]
Ok, feel free to try connecting the app again, and let me know if you're successful.

**Comment by LucBerge** [comment]
It's working ! PR for actions coming soon

**Comment by LucBerge** [comment]
@dannyroosevelt Btw, new endpoints are available 24/7 now.., Better late than never.

Do you need help to translate the contract to go production?

**Comment by dannyroosevelt** [comment]
> Would it make sense to have one enedis app for the sandbox mode and one for production ?

Maybe it can be handled within a single app -- what's the difference, just the host? 

-- 

>Do you need help to translate the contract to go production?

I'm checking with the team on how to handle next steps on the contract

**Comment by LucBerge** [comment]
> what's the difference, just the host?

No.

Sandbox:
- Call the token request endpoint (with my own `cleint_id` and `secret_id`-> Redirect to pipedream when done
- 
Production:
- Call the Authorize endpoint -> Redirect to pipedream when done
- Call the token request endpoint (with Pipedream's `client_id` and `secret_id`-> Redirect to pipedream when done

-------

There are differences for the actions as well:

Sandbox:
- Call the sandbox endpoints
- The `usage_point_id` is known by the user

Production:
- Call the production endpoints
- The `usage_point_id` ARE known when authorize request is complete. The user must select the available usage point from a dropdown list

-------

Actions in sandbox mode are done but I can't verify two of them. Two endpoints return an `error 500` : `Technical error. Please try later.` since three days...
-	https://ext.hml.api.enedis.fr/metering_data_dp/v5/daily_production
-	https://ext.hml.api.enedis.fr/metering_data_plc/v5/production_load_curve

I'm waiting for a `200 status` before PR

**Comment by LucBerge** [comment]
Ok they told me by email that the usage pointI use (my house) does not have production enabled so it's normal... I can use the following virtual usage point instead:
- 24880057139941
- 10284856584123

**Comment by LucBerge** [comment]
@dannyroosevelt The merge automatically closed this issue, while it should not be closed. The app is running in sandbox mode not production yet

**Comment by vunguyenhung** [comment]
Thank you for informing us @LucBerge! I've reopened the issue, though I will also move this in the Blocked column until this is deployed to prod

**Comment by LucBerge** [comment]
Why do you consider it as blocked?

**Comment by LucBerge** [comment]
> @LucBerge unfortunately the contract is still on hold due to the language barrier.

@dannyroosevelt How can I help ?

**Comment by dannyroosevelt** [comment]
It's on the backlog to address but we're a small team and are focused on other priorities right now.

**Comment by LucBerge** [comment]
Sure, no problem 😄

**Comment by sergio-eliot-rodriguez** [comment]
https://pipedream.com/apps/enedis

**Comment by LucBerge** [comment]
@sergio-eliot-rodriguez why did you move it from **Blocked** to **Done**? Why is the issue closed?

**Comment by sergio-eliot-rodriguez** [comment]
@LucBerge , I see that Enedis has been integrated -- https://pipedream.com/apps/enedis
was the wrong application integrated?

**Comment by LucBerge** [comment]
> @LucBerge , I see that Enedis has been integrated -- https://pipedream.com/apps/enedis was the wrong application integrated?

This ticket is not only about app integration. The app is in sandbox mode and need to go to production. In order to go to production, pipedream need to sign a contract with Enedis. This is the reason why it is tagged as `blocked`.

Please read the full thread.
